### DISCUSSION TITLE
Extract out a signal from a store

### DISCUSSION THREAD

**Discussion by quazar-omega** [solidjs/solid#1775] [Q&A]
Hi, I was playing around with the store functionality and I wanted to be able to update the state of a specific part of it from within a child component.  

The structure describes an array of lists that contain some items, all of them are identified by an id that I generate with `randomUUID()` since [solid-dnd](https://github.com/thisbeyond/solid-dnd/), which I'm using for drag and drop behavior, requires that draggables and droppables have a unique id
<details>
<summary>
The structure is this:
</summary>

```tsx
const [lists, setLists] = createStore([
	{
		id: String(self.crypto.randomUUID()),
		name: 'A',
		items: [
			{ id: String(self.crypto.randomUUID()), name: 'a' },
			{ id: String(self.crypto.randomUUID()), name: 'b' },
			{ id: String(self.crypto.randomUUID()), name: 'c' }
		]
	},
	{
		id: String(self.crypto.randomUUID()),
		name: 'B',
		items: [
			{ id: String(self.crypto.randomUUID()), name: 'd' },
			{ id: String(self.crypto.randomUUID()), name: 'e' },
			{ id: String(self.crypto.randomUUID()), name: 'f' }
		]
	},
])
```
</details>

I generate the lists in this way:
```ts
<For each={lists}>
	{(list) => <ItemList id={list.id} name={list.name} items={list.items} setItems={setLists} />}
</For>
```

I pass the `setLists` setter to the lists so that I can create new items from within the `ItemList` component,   
but I'm having an issue at this point, this button in `ItemList` calls on the function that makes a new item:
```tsx
<button onClick={() => {
	newItem(prompt('Give the item a name'))
}}>
	Add a new element
</button>
```

In order to add a new element I have to address that specific list first, this works as an example, it will update the list at index 0:
```tsx
function newItem(itemName: string) {
	props.setItems(
		0,
		'items',
		(items) => {
			return [
				...items,
				{
					id: String(self.crypto.randomUUID()),
					name: itemName
				}
			]
		}
	)
}
```
but I need its index in the store to have it add an item to the appropriate list, I thought of passing the `lists` getter too along with setter to the `ItemList`, so I could go and get the index given the id of the `ItemList` stored in `props.id`, and indeed it works by adding the property in:

(in the parent component)
```tsx
<ItemList id={list.id} name={list.name} items={list.items} setItems={setLists} getLists={lists} />
```
(in `ItemList`)
```tsx
function getListIndex(id: Id) {
	return props.getLists.map(
		(list) => {
			return list.id
		}).indexOf(id)
}

function newItem(itemName: string) {
	props.setItems(
		getListIndex(props.id),
		'items',
		(items) => {
			return [
				...items,
				{
					id: String(self.crypto.randomUUID()),
					name: itemName
				}
			]
		}
	)
}
```

But I find it quite ugly, so I was wondering if it is possible to extract the specific setter of an element that is inside a store, to be able to pass it out and use directly, I imagined that something like this:
```tsx
for (const listIndex in lists) {
	console.log(setLists[listIndex])
}
```
would output the setters for each list, since it says in the [documentation](https://docs.solidjs.com/references/api-reference/stores/using-stores) that stores are trees of signals, but it just gives `undefined` .

Maybe there's an even more obvious way to do this that I haven't considered, do let me know if that's the case as I'm a complete novice with the framework.
Thank you!

**Comment by quazar-omega** [comment]
Found what I was looking for in [this comment](https://github.com/solidjs-community/solid-primitives/pull/452#issuecomment-1566677664), thank you @thetarnav <3.
So in the end it simply became:

(parent component)
```tsx
<For each={lists}>
	{(list) => <ItemList id={list.id} name={list.name} items={list.items} />}
</For>
```
(child component `ItemList`)
```tsx
const [items, setItems] = createStore(props.items)

function newItem(itemName: string) {
	setItems(
		(items) => {
			return [
				...items,
				{
					id: String(self.crypto.randomUUID()),
					name: itemName
				}
			]
		}
	)
}
```

**Reply by quazar-omega** [reply]
Actually I verified that the changes I make on this new store do not apply to the original, I don't understand why

**Reply by thetarnav** [reply]
I think that since you’re destructuring the array, you can’t rely on it’s reference in the original store.
Try creating a “scoped” store from a stable reference, or update the array by mutating it.

**Reply by quazar-omega** [reply]
Oh thanks for answering!!
> Try creating a “scoped” store from a stable reference

I was doing that just now, and was literally about to post my results when I saw your comment lol, I'll tag you once I post that.

> or update the array by mutating it

Could you please give me an example, there's a way to mutate the object in a store manually?

**Reply by thetarnav** [reply]
https://github.com/solidjs/solid/discussions/866

**Reply by quazar-omega** [reply]
I saw that one, isn't the first option what I'm doing already?

**Comment by quazar-omega** [comment]
@thetarnav A new solution that is working flawlessly is this, but I still dislike how it looks since I'm referring to the getter and setter functions as elements of an array.  
I'm sure there's some Javascript-fu that could fix this one last hurdle, but nothing comes to mind, I tried to destructure the store into an array and saving its two elements in an object to get named properties but it seems the references are lost that way.

```tsx
<For each={lists}>
	{(list) => <ItemList id={list.id} name={list.name} items={createStore(list.items)} />}
</For>
```

```tsx
function newItem(itemName: string) {
	props.items[1](
		(items) => {
			return [
				...items,
				{
					id: String(self.crypto.randomUUID()),
					name: itemName
				}
			]
		}
	)
}
```
